05_debug/debug_module.c

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/ctype.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/printk.h>
+
+#include <linux/sysfs.h>
+#include <linux/proc_fs.h>
+#include <linux/device.h>
+#include <linux/uaccess.h>
+
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+MODULE_AUTHOR("Andrii Synenko");
+MODULE_DESCRIPTION("List data module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+
+#define PROCFS_BUFF_SIZE 255
+#define PROCFS_DIR_NAME	"debug_module"
+#define PROCFS_ENT_NAME	"upper"
+#define SYSFS_DIR_NAME	PROCFS_DIR_NAME
+#define SYSFS_ENT_NAME	"lower"
+#define STAT_ENT_NAME	"stat"
+
+struct convertor_struct {
+	char *str_buff;
+
+	int converted_last;
+	int converted_total;
+	int processed_last;
+	int processed_total;
+
+} sysfs_conv, procfs_conv;
+
+static int convert_case(bool to_lover, struct convertor_struct *conv,
+						ssize_t str_size)
+{
+	size_t indx = 0;
+
+	conv->converted_last = 0;
+
+	if (conv == NULL || conv->str_buff == NULL) {
+		pr_err("%s: convertor or buffer is NULL\n", __func__);
+		return -EINVAL;
+	}
+
+	while (conv->str_buff[indx] != '\0' &&
+		conv->str_buff[indx] != '\n' &&
+		indx < str_size) {
+		/*
+		 *!( isupper XOR to_lover )
+		 * |   t      |    t      | t - lovercase output
+		 * |   f      |    f      | t - uppercase output
+		 */
+		if (!(isupper(conv->str_buff[indx]) ^ to_lover)) {
+			conv->str_buff[indx] = (to_lover) ?
+					tolower(conv->str_buff[indx]) :
+					toupper(conv->str_buff[indx]);
+			conv->converted_last++;
+		}
+		indx++;
+	}
+	conv->processed_last  = indx;
+
+	conv->converted_total += conv->converted_last;
+	conv->processed_total += conv->processed_last;
+
+	return 0;
+}
+
+int print_stat(char *buf, struct convertor_struct *conv)
+{
+	if (conv == NULL || buf == NULL) {
+		pr_err("%s: convertor or buffer is NULL\n", __func__);
+		return -EINVAL;
+	}
+
+	return sprintf(buf, "processed last : %d char(s)\n"
+			"converted last : %d char(s)\n"
+			"converted total: %d char(s)\n"
+			"processed total: %d char(s)\n",
```

Comment:
Different order of processed/converted for last/total is inconvenient to read.

Reply:
Fixed